### ISSUE TITLE
Feature request: dynamic keep-alive

### ISSUE THREAD

**Issue by sshock** [ngtcp2/ngtcp2#1225]
Hi, I would like to use something like a 30 second keep-alive and a 40 second idle timeout.

However, I see that keep-alive PINGs are not retransmitted, but rather simply transmitted at the specified rate.

Which means a 30s keep-alive and 40s idle timeout doesn't make sense, since a single lost keep-alive packet would terminate the connection.

With the current design, I need to either reduce the keep-alive to 15 seconds, or increase the idle timeout to 95 seconds, if I want to ensure at least 3 keep-alive packets must be lost before it will idle timeout.

However, if ngtcp2 could support two keep-alive timeouts, one for normal times and (a faster) one for when they are unACKed, then I could have a regular keep-alive of 30s and a fast keep-alive of 2s, with an idle timeout of 40s.  There would have to be 5 lost packets (at 30, 32, 34, 36, and 38 seconds) before it would idle timeout.

That would give us the best of both worlds (the ability to have a higher regular keep-alive, to reduce load, while also being able to have a lower idle timeout, to detect connection down more quickly).

**Comment by tatsuhiro-t** [comment]
I am wondering this could be achieved by just triggering PTO probe for keep-alive PING only packets.
PTO is much shorter than keep-alive interval, and keep-alive packet itself is not going to retransmit, but PTO probes will be sent to elicit acknowledgement from a remote endpoint, which effectively does keep-alive job.

**Comment by sshock** [comment]
Yeah, I think it makes sense to trigger PTO probes for keep-alive PING packets.  This would solve the problem nicely.  And if there is any concern about this change in behavior, you could make it optional.

Out of curiosity, what exactly would go in the PTO probes?  Wouldn't they just contain like a PING frame, and if so, how would they differ from just retransmitting the keep-alive packet?

**Comment by tatsuhiro-t** [comment]
PTO probe can be any packet that contains ack eliciting frames.  If nothing to be sent, PING is included.
We usually do not retransmit meaningless packet that only contains PING even if it itself has some purpose.

**Comment by sshock** [comment]
Ok, that makes sense.  So even though these PTO probe packets may end up looking a lot like the keep-alive packets, they aren't retransmissions, and I guess that also means they each have a different packet number.

With this approach, how many probes would be sent after an un-ACKed keep-alive packet?  Do they start out fast then have exponential back-off?  Would they keep going until the next regular (e.g., 30s) keep-alive goes out?

**Comment by tatsuhiro-t** [comment]
https://datatracker.ietf.org/doc/html/rfc9002#name-sending-probe-packets describes how PTO probes work.

**Comment by sshock** [comment]
Thanks.  I had already skimmed that section earlier, but hadn't read through it thoroughly.  I see "Consecutive PTO periods increase exponentially", which answers my first question, and matches what I have seen with lost stream data.

Unfortunately, I'm still not 100% sure I understand whether your proposed idea would end up looking like A or B in the scenario where the peer stops responding:

**A:** 
1. Keep-alive packet sent after keep-alive timeout (e.g., 30s)
2. PTO probe sent after PTO timeout (e.g., 50ms)
3. PTO probes continue to be sent (e.g., after 100ms, 200ms, 400ms, etc.) until it would have been time for the next keep-alive (e.g., 30s after first keep-alive).
4. Send the next keep-alive packet and reset the PTO (e.g., back to 50ms)
5. GOTO 2

**B:** 
1. Keep-alive packet sent after keep-alive timeout (e.g., 30s)
2. PTO probe sent after PTO timeout (e.g., 50ms)
3. PTO probes continue to be sent (e.g., after 100ms, 200ms, 400ms, etc.) forever (until idle timeout close)

**Comment by tatsuhiro-t** [comment]
PTO probe packets are sent until a packet which contains ACK frame is received.  If they keeps lost, then connection is unstable and unusable.
Keep alive packet will be sent according to the connection state.  It might be sent as PTO probe.

**Comment by sshock** [comment]
Thanks.  I think I understand that if we decide to send keep-alive as PTO probe, after conn is idle for keep-alive timeout, we'll send a PTO probe packet for the keep-alive, and then continue to send PTO probes (until we get an ACK or we close due to idle timeout).

The part I don't understand is what happens after another keep-alive timeout period has elapsed.  The current design sends another keep-alive packet, so with the new design would we start a whole new series of PTO probes going?

Anyway, sorry for bugging you with so many questions.  I can just wait and test it out after you make the change (if you decide to change it).

**Comment by tatsuhiro-t** [comment]
> The part I don't understand is what happens after another keep-alive timeout period has elapsed. The current design sends another keep-alive packet, so with the new design would we start a whole new series of PTO probes going?

If the interval of PTO probe is longer than keep-alive, then keep-alive packet will be sent.  There is just one PTO in a connection.  The existing PTO continues.

**Comment by sshock** [comment]
Thanks for making this change and for being patient with me!

**Comment by sshock** [comment]
Good news.  I tested this out today, and it worked wonderfully!

Using a 30s keep-alive and no idle timeout, I forced the network to go down shortly after connecting.  A keep-alive was sent after 30s, followed by PTO probes that started out quickly but increased exponentially.  Once the PTO probes were more than 30s apart, I saw keep-alives going out at regular 30s intervals.

This is exactly what I expected to happen after reading Tatsuhiro's last comment.
